### DISCUSSION TITLE
Is socks5h supported?

### DISCUSSION THREAD

**Discussion by Klaaktu** [shadowsocks/shadowsocks-rust#441] [Q&A] [answered]
Hi. Is socks5h(remote DNS resolution) supported? I see that there are commits about DNS relay, but README only mentions socks5.
I have set socks5h as global proxy in Ubuntu, but often websites wait for some time before loading, as if host names are being resolved locally (or because of something else, I'm not sure). Also `apt` only supports socks5h.

**Comment by zonyitoo** [comment]
It depends on your client's configuration. All socks5 proxies supports remote domain name resolution.

**Comment by database64128** [comment, marked as answer]
Some applications don't recognize `socks5h` and simply treat it as an HTTP proxy, which can cause issues. For example, if you only set `ALL_PROXY=socks5h://localhost:1080/`, Git LFS will treat it as an HTTP proxy. But if you only set `ALL_PROXY=socks://localhost:1080/`, cURL will use your local resolver, which may not be desirable in your network environment.

In practice, you can set it this way:

- `ALL_PROXY=socks://localhost:1080/`
- `HTTP_PROXY=http://localhost:8080/`
- `HTTPS_PROXY=http://localhost:8080/`

Most applications (Git LFS, cURL, etc) will use the HTTP proxy and work correctly. Applications that support `ALL_PROXY` will recognize `socks` and most likely send requests directly without your local resolver getting involved.

**Comment by Klaaktu** [comment]
I see. Thanks for the answers guys.

p.s. It seems Ubuntu network settings gui doesn't support specifying scheme. I set `socks5h://127.0.0.1:1080/` in the `socks` field and this is what the environment variable looks like:
```
$ echo $all_proxy
socks://socks5h://127.0.0.1:1080/
```

**Reply by zonyitoo** [reply]
If you are looking for a way to set proxy with environemnt variables, why not using HTTP mode?

**Reply by Klaaktu** [reply]
Yep I have both HTTP and Socks now:
```
$ echo $http_proxy
http://127.0.0.1:8388/
$ echo $all_proxy
socks://127.0.0.1:1080/
```
